hail/src/main/scala/is/hail/expr/types/virtual/TUnion.scala

```diff
@@ -0,0 +1,145 @@
+package is.hail.expr.types.virtual
+
+import is.hail.annotations.ExtendedOrdering
+import is.hail.expr.ir.IRParser
+import is.hail.expr.types.physical.PType
+import is.hail.utils._
+import org.apache.spark.sql.Row
+import org.json4s.CustomSerializer
+import org.json4s.JsonAST.JString
+
+import scala.reflect.ClassTag
+
+final case class Case(name: String, typ: Type, index: Int) {
+
+  def unify(cf: Case): Boolean =
+    name == cf.name &&
+      typ.unify(cf.typ) &&
+      index == cf.index
+
+  def pretty(sb: StringBuilder, indent: Int, compact: Boolean) {
+    if (compact) {
+      sb.append(prettyIdentifier(name))
+      sb.append(":")
+    } else {
+      sb.append(" " * indent)
+      sb.append(prettyIdentifier(name))
+      sb.append(": ")
+    }
+    typ.pretty(sb, indent, compact)
+  }
+}
+
+class TUnionSerializer extends CustomSerializer[TUnion](format => (
+  { case JString(s) => IRParser.parseUnionType(s) },
+  { case t: TUnion => JString(t.parsableString()) }))
+
+object TUnion {
+  private val requiredEmpty = TUnion(Array.empty[Case], true)
+  private val optionalEmpty = TUnion(Array.empty[Case], false)
+
+  def empty(required: Boolean = false): TUnion = if (required) requiredEmpty else optionalEmpty
+
+  def apply(required: Boolean, args: (String, Type)*): TUnion =
+    TUnion(args
+      .iterator
+      .zipWithIndex
+      .map { case ((n, t), i) => Case(n, t, i) }
+      .toArray,
+      required)
+
+  def apply(args: (String, Type)*): TUnion =
+    apply(false, args: _*)
+}
+
+final case class TUnion(cases: IndexedSeq[Case], override val required: Boolean = false) extends Type {
+  lazy val physicalType: PType = ???
+
+  lazy val types: Array[Type] = cases.map(_.typ).toArray
+
+  lazy val caseIdx: collection.Map[String, Int] = toMapFast(cases)(_.name, _.index)
+
+  lazy val caseNames: Array[String] = cases.map(_.name).toArray
+
+  def size: Int = cases.length
+
+  override def unify(concrete: Type): Boolean = concrete match {
+    case TUnion(cfields, _) =>
+      cases.length == cfields.length &&
+        (cases, cfields).zipped.forall { case (f, cf) =>
+          f.unify(cf)
+        }
+    case _ => false
+  }
+
+  override def subst(): TUnion = TUnion(cases.map(f => f.copy(typ = f.typ.subst())))
+
+  def index(str: String): Option[Int] = caseIdx.get(str)
+
+  def selfCase(name: String): Option[Case] = caseIdx.get(name).map(i => cases(i))
+
+  def hasCase(name: String): Boolean = caseIdx.contains(name)
+
+  def getCase(name: String): Case = cases(caseIdx(name))
+
+  def fieldType(name: String): Type = types(caseIdx(name))
+
+  def rename(m: Map[String, String]): TUnion = {
+    val newFieldsBuilder = new ArrayBuilder[(String, Type)]()
```

Comment:
Do you need to worry about duplicate names after the rename?

Reply:
Hmm, good question.  This was copied unmodified from `TStruct`.  I created an issue to follow up on this: https://github.com/hail-is/hail/issues/6409